### ISSUE TITLE
[V2] Android Exception java.util.NoSuchElementException StackController.java:228

### ISSUE THREAD

**Issue by aminosman** [wix/react-native-navigation#5117]
### Issue Description

Seeing this error frequently in production but unable to reproduce. Any thoughts on what we could be doing to cause this / if there is a way to prevent this?

### Steps to Reproduce / Code Snippets / Screenshots

```  
  at java.util.ArrayList$Itr.next (ArrayList.java:863)
  at com.reactnativenavigation.viewcontrollers.stack.StackController.removeChildrenBellowTop (StackController.java:228)
  at com.reactnativenavigation.viewcontrollers.stack.StackController.access$000 (StackController.java:36)
  at com.reactnativenavigation.viewcontrollers.stack.StackController$2.onSuccess (StackController.java:209)
  at com.reactnativenavigation.viewcontrollers.stack.StackController.lambda$push$4$StackController (StackController.java:174)
  at com.reactnativenavigation.viewcontrollers.stack.StackController$$Lambda$4.run (Unknown Source:8)
  at com.reactnativenavigation.anim.NavigationAnimator$1.onAnimationEnd (NavigationAnimator.java:65)
  at android.animation.Animator$AnimatorListener.onAnimationEnd (Animator.java:552)
  at android.animation.AnimatorSet.endAnimation (AnimatorSet.java:1294)
  at android.animation.AnimatorSet.doAnimationFrame (AnimatorSet.java:1079)
  at android.animation.AnimationHandler.doAnimationFrame (AnimationHandler.java:146)
  at android.animation.AnimationHandler.access$100 (AnimationHandler.java:37)
  at android.animation.AnimationHandler$1.doFrame (AnimationHandler.java:54)
  at android.view.Choreographer$CallbackRecord.run (Choreographer.java:947)
  at android.view.Choreographer.doCallbacks (Choreographer.java:761)
  at android.view.Choreographer.doFrame (Choreographer.java:693)
  at android.view.Choreographer$FrameDisplayEventReceiver.run (Choreographer.java:935)
  at android.os.Handler.handleCallback (Handler.java:873)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:7032)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:494)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:965)
 ```

---
### Environment
* React Native Navigation version: 2.18.5
* React Native version: 0.57.8
* Platform(s) (iOS, Android, or both?): Android
* Device info (Simulator/Device? OS version? Debug/Release?): Device Android 7.1-9 Release

**Comment by ball-hayden** [comment]
We've seen it while trying to navigate to `setStackRoot` to be the same as it is already.

i.e.

```js
setStackRoot({
  id: 'uk.co.playerdata.About',
  name: 'uk.co.playerdata.About',
  options: {
    topBar: {
      title: { text: 'About' },
      leftButtons: [
        {
          id: 'leftNav',
          icon: {
            uri: 'file:///data/user/0/uk.co.playerdata.mobile/cache/1pi43jj_30@2.625x.png',
            scale: 2.625,
          },
        },
      ],
    },
    sideMenu: { left: { enabled: true } },
  },
});

// works as intended

setStackRoot({
  id: 'uk.co.playerdata.About',
  name: 'uk.co.playerdata.About',
  options: {
    topBar: {
      title: { text: 'About' },
      leftButtons: [
        {
          id: 'leftNav',
          icon: {
            uri: 'file:///data/user/0/uk.co.playerdata.mobile/cache/1pi43jj_30@2.625x.png',
            scale: 2.625,
          },
        },
      ],
    },
    sideMenu: { left: { enabled: true } },
  },
});

// crashes, as above.

```

**Comment by ball-hayden** [comment]
Stepping through it, the problem seems to be that `setRoot` is pushing the new component before attempting to remove old components.

https://github.com/wix/react-native-navigation/blob/0720bda67973acd96a262fbde7c2abb9fb8b7848/lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/StackController.java#L186-L191

In the case that the old and new component have the same `id`, we end up with the `IdStack` having a single entry in its `map` for the component, but two entries in its `deque` list.

The size of the stack is based on the `deque`, so our stack has a size of 2:

https://github.com/wix/react-native-navigation/blob/0720bda67973acd96a262fbde7c2abb9fb8b7848/lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/IdStack.java#L49-L51

Both entries in the stack are the top (as they both have the same ID), so we don't remove either of them...

https://github.com/wix/react-native-navigation/blob/0720bda67973acd96a262fbde7c2abb9fb8b7848/lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/StackController.java#L213-L222

...leaving us with a stack that still has a size of 2 (greater than 1).

The iterator then increments past both entries, leaving us with our `NoSuchElementException`.

**Comment by ball-hayden** [comment]
@aminosman I'd suggest your issue might be either a duplicate ID, or you're trying to navigate to the same place twice.

My question is what is the correct behaviour here.
* Should `setRoot` create a new view and dispose of the old one?
* Should `setRoot` notice this condition and do nothing?
* Should `setRoot` notice this condition and raise a clear exception (so I, as a developer, know this is the wrong thing to be doing)?

I'd lean towards the first option here, but it should probably be consistent with iOS behaviour (which doesn't crash, FWIW). I'm afraid my Objective-C isn't good enough to work out what the current iOS behaviour is, but I'm guessing it's disposing of the old view and creating a new one?

**Comment by ball-hayden** [comment]
Vaguely related to https://github.com/wix/react-native-navigation/issues/4333, although that's the iOS equivalent I think?

**Comment by aminosman** [comment]
@ball-hayden I appreciate all the digging you did, I was generating random IDs for the screens in the new stack as this was an issue but as they are time based I saw how you could get 2 of the same IDs somehow two pushes happen in the same millisecond. After digging I see in some of the logs android seems to be executing the code multiple times as I see our app events fired in triplicate at times. I am not sure what is causing this but I have added debouncing to the push and will see if this alleviates the issue in the mean time. I do agree that setRoot should destroy the old components.

**Comment by guyca** [comment]
Hey guys, this should be fixed in `2.20.2-snapshot.355`
